File: Java/Labs/Module 2/Lab 2/Q2 - Lego Blocks/Main.java
import java.util.*;

class Result{
    public static int legoBlocks(int a,int b){
        if(a<2||b<1) return 0;
        if(b==1) return 1;
        long[] c=new long[b+1];
        for(int i=0;i<c.length;i++){
            c[i]=-1;
        }
        d(c,b);
        for(int i=0;i<c.length;i++){
            long e=1;
            for(int j=0;j<a;j++){
                e=(e*c[i])%MOD;
            }
            c[i]=e;
        }
        long[] f=new long[b+1];
        for(int i=0;i<f.length;i++){
            f[i]=-1;
        }
        g(c,f,b);
        return (int)(f[b]%MOD);
    }
    static long MOD=1000000000+7;
    static long g(long[] c,long[] f,int i){
        if(f[i]==-1){
            if(i==1){
                f[i]=1;
            }
            else{
                f[i]=c[i];
                for(int j=1;j<i;j++){
                    f[i]-=(g(c,f,j)*c[i-j])%MOD;
                }
            }
        }
        return f[i];
    }
    static long d(long[] c,int i){
        if (i<0) return 0;
        if(c[i]==-1){
            if(i==0||i==1)
                c[i]=1;
            else
                c[i]=(d(c,i-1)+d(c,i-2)+d(c,i-3)+d(c,i-4))%MOD;
        }
        return c[i];
    }
}
public class Main{
    public static void main(String[] args){
        Scanner s=new Scanner(System.in);
        int n=s.nextInt();
        int m=s.nextInt();
        System.out.print(Result.legoBlocks(n,m));
    }
}
                                                                                                                            

File: C++/Labs/Module 2/Lab 2/Q2 - Lego Blocks/main.cpp
#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

int power(int base,int exponent,int modulus){
    if(exponent==0) return 1;
    long long result =1;
    long long x=base;
    while(exponent>0){
        if(exponent%2==1){
            result=(result*x)%modulus;
        }
        x=(x*x)%modulus;
        exponent/=2;
    }
    return static_cast<int>(result);
}

int legoBlocks(int n,int m){
    const int MOD=1000000007;
    vector<int> row_combinations={1,1,2,4};
    while(row_combinations.size()<=m){
        int sum_last_four=0;
        for(int i=row_combinations.size()-4;i<row_combinations.size();++i){
            sum_last_four=(sum_last_four+row_combinations[i])%MOD;
        }
        row_combinations.push_back(sum_last_four);
    }
    vector<int> total;
    for(int c:row_combinations){
        total.push_back(power(c,n,MOD));
    }
    vector<int> unstable={0,0};
    for(int i=2;i<=m;++i){
        auto f=[&](int j){return(total[j]-unstable[j])*total[i-j]%MOD;};
        int result=0;
        for(int j=1;j<i;++j){
            result=(result+f(j))%MOD;
        }
        unstable.push_back(result);
    }
    return(total[m]-unstable[m]+MOD)%MOD;
}
int main(){
    int n,m;
    cin>>n>>m;
    cout<<legoBlocks(n,m)<<endl;
    return 0;
}
                                                                                                                            